File: 345_Reverse_Vowels_of_a_String.java
/*
https://leetcode.com/problems/reverse-vowels-of-a-string/
Write a function that takes a string as input and reverse only the vowels of a string.

Use pointers to swap the first vowel on the left and right, then increment pointers inwards until overlap.
Time - O(n)
Space - O(n)
*/

public class Solution {
    public String reverseVowels(String s) {

        char[] sArray = s.toCharArray();
        Set<Character> vowels = new HashSet<Character>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
        int left = 0;
        char temp;
        int right = s.length() - 1;

        while (left < right) {
            while (left < s.length() && !vowels.contains(s.charAt(left))) {
                left++;
            }
            while (right >= 0 && !vowels.contains(s.charAt(right))) {
                right--;
            }
            temp = s.charAt(left);
            sArray[left] = sArray[right];
            sArray[right] = temp;
            left++;
            right--;
        }

        return new String(sArray);

    }
}

File: 345_Reverse_Vowels_of_a_String.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/reverse-vowels-of-a-string/
# Write a function that takes a string as input and reverse only the vowels of a string.

# Find list of indices of vowels in s. Swap vowel at first index with vowel at last index, until meeting in middle
# of list of vowels.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowels = {"a", "e", "i", "o", "u"}
        vowels |= {c.upper() for c in vowels}

        vowel_i = [i for i, c in enumerate(s) if c in vowels]
        n_vowel = len(vowel_i)

        s = [c for c in s]
        for j in range(n_vowel // 2):
            s[vowel_i[j]], s[vowel_i[n_vowel - j - 1]] = s[vowel_i[n_vowel - j - 1]], s[vowel_i[j]]

        return "".join(s)